internal/runner/operation_deploy.go

```diff
@@ -24,6 +24,19 @@ func (r *Runner) executeDeployOp(
 		panic("operation not expected type")
 	}
 
+	if op.Deploy.Artifact == nil {
```

Comment:
Not sure of the consequences of this change, but the reasoning behind this is Pipelines queue all of its Step operations at once ahead of time. That means we can't attach the "latest" build artifact then because it won't exist if there is a Build step prior that hasn't run yet. Pipelines will set the artifact to nil on the job to be queued and then the deploy can look it up here.

Reply:
I think doing this also gets us one step closer to being able to update the `DeployOp` to use a `Ref_Artifact` instead of requiring the full artifact proto for a deploy:

https://github.com/hashicorp/waypoint/blob/d557c0a5f5b9fe8fad599d6fad8ea11b97da4173/pkg/server/proto/server.proto#L1965-L1968

https://github.com/hashicorp/waypoint/issues/2884

Reply:
[Question] If an earlier build step in a pipeline fails, will the pipeline stop in its tracks (i.e. no remaining queued steps are run), or will it continue on failure? 

I'm thinking that if the pipeline continues to run, then what could possibly happen is with an earlier build phase failure, the pipeline could then unexpectedly re-deploy an old artifact by looking up the "latest" artifact here, since the one that was just attempted to be built failed (in this scenario).

Setting that question aside, I think this is a good way to solve the problem!

Reply:
Good question! This is definitely and edge case I think, but possible if you tried hard enough with how pipelines are right now.

All steps are translated by Waypoint into jobs that run through the job system. By default, steps depend on the previous step in the pipeline unless it's the "root" or first step. So if that build step fails, the downstream jobs will be cancelled due to the job failure. You'd need to set the `depends_on` field in a weird way where the deploy relies on a successful step prior to the build failure for that to possibly happen but it wouldn't be the default case.

If we end up implementing #2884 too we can make a Deploy step include a `Ref_Artifact` that might reference a prior pipeline step and combined with explicitly having your deploy step `depends_on` the build, this might help in this case. 🤔